### ISSUE TITLE
[feat]: line chart type renders one chart per series instead of overlaying multiple series with a shared legend

### ISSUE THREAD

**Issue by mayura-andrew** [goptics/vizb#336]
### Problem / use case

We're using vizb line in "solo" mode (--select step,colN{label}, one point per step, not aggregated) to render multiple time-series lines that share the same unit/axis — e.g. train loss and validation loss, which we want on one chart so it's easy to see when they diverge.

The generated DataSet JSON correctly groups all series under one dataset (confirmed — one "loss" dataset with N distinct "type" values per xAxis entry). But vizb ui -c line renders one chart per series rather than one chart with all series overlaid and a shared, clickable legend (like [this ECharts stacked-line example](https://echarts.apache.org/examples/en/editor.html?c=line-stack) — legends are clustered and clickable to toggle a line's visibility).

We checked --stack on vizb line against this data shape first, in case it already covered this:

```
$ vizb line test.csv -o test.json -n loss --select "step,col0{Alpha}" --select "step,col1{Beta}" --stack
flag "stack" skipped: requires axis "y" (present: [x])
```
It only applies to the "grouped" axis mode, not "solo" mode.

Would it be possible for vizb ui -c line to support rendering multiple series from one dataset as a single chart with an overlaid multi-line + legend view (ECharts natively supports legend-click-to-toggle when configured this way), at least as an option? Happy to provide more detail on our exact use case if useful — this is for [gomlx](https://github.com/gomlx/gomlx)'s gomlx_checkpoints --plot, comparing training-metric curves.

Let me know if you want any wording changes before you send it.

**Comment by janpfeifer** [comment]
Another [E-Chart example](https://echarts.apache.org/examples/en/editor.html?c=line-stack&theme=dark&code=nZO7buMwEEX7fAWbhRoXlugAkRdbGNjsIkBS5VUELsbSWCZCkwZFGTIC_3tmaJFxHnDhhhKvLnnuDCm78coa8Ue8XQjhldc4Da80wd5PRXbvoXrFWtwqgxl92I_YaK32apOsTjUNOnJDr9rk0tigqaOpBg9T8ZJdr0HpbCSyR8PkWd3y5EnVaOPkr3JYeX67R3DVSlybhvFz3lUI6xQazrZF51UFtFvQKcWK5Xy86QfJWwp5EGKqxqmUSeOSF8hfX3aYRGFhvbfrY0tljQdlbmGBekqVd3jclIXt495LBN-51E4hWtjirL1ZQ8PiPqg8hsX9jDqX-rnbkCWrwGNj3S5l6UwNbvcfqKYl6BYPemzsnTXcsocO-fGMdZitOn78cyq0Ew5d7Uw2j-TdD-QtaNolOlqkhpPlJVhiOQbWbB2OcxCH9ZpPK2ot3yASH6xPR5VS58V4JHJZ0DDO-W0yEiVJheQhH3PMIcY39NEFOhdfBPwV40vCF4wvmC-ZL0_zj-7s2eVfhlqJX4TyL4mfJ_7kND_-JufCJRcvGS4ZLrl4-VF8cRL--c88O8MVZyg5Q8kZSs6QhxPIQwq6GjEGjfOL_e93&enc=deflate) of stacked plots in the same figure, with a legend on the top-right for visibility.

**Comment by fahimfaisaal** [comment]
Thanks for the clear write-up and the ECharts examples, that helps a lot.

If I’m reading the use case right, you have something like wide step-series data (one row per step, one column per curve):

```csv
step,col0,col1
1,0.52,0.61
2,0.48,0.58
```

and you’re using solo multi `--select`:

```
vizb line test.csv -n loss \
  --select "step,col0{Alpha}" \
  --select "step,col1{Beta}"
```

That path is multi-stat mode: one dataset, but the UI intentionally splits charts by stat/type (one chart per series). So the JSON looking “merged” while the line UI still shows separate charts is consistent with how that mode works today, not quite the multi-line overlay you’re after.

I don’t have your exact test.csv layout, but if the goal is several series on one chart + shared legend, group (with series on the Y dimension) should already cover it.

If the file is wide (column per series), try putting step on X and column names on Y via `--col-axis`:

```bash
vizb line test.csv -n loss -g step -p x --col-axis y --select "col0{Alpha},col1{Beta}" -o loss.html
```
Output:

<img width="1273" height="726" alt="Image" src="https://github.com/user-attachments/assets/d46124fb-3c1e-4a8a-9dcb-679b006bbc89" />

With `--stack`:

<img width="1273" height="726" alt="Image" src="https://github.com/user-attachments/assets/c985850b-ef24-41fa-a77e-317bc8ef8eb2" />

That pattern is the same idea as the concurrency example: column names become in-chart series on one chart instead of one chart per column.

If you reshape to long/tidy form:

```csv
step,series,value
1,Alpha,0.52
1,Beta,0.61
```

then:

```bash
vizb line test.csv -n loss -g series,step -p y,x --select value -o loss.html
```

Here series -> Y (legend series), step -> X.

On `--stack`

`--stack` on line is a different visual than “multiple independent lines with a legend.” It’s aimed at grouped 2D data and renders as a stacked area chart (part-to-whole / contribution to the total per X), in the spirit of the [ECharts area-stack](https://echarts.apache.org/examples/en/editor.html?c=area-stack) example - not a plain multi-line overlay.

The doc ref: https://vizb.goptics.org/charts/line/#2d-x--y-axes

Group aggregates rows that share the same key (sum for CSV/JSON). If you already have one value per (step, series), that’s usually fine.

**Comment by mayura-andrew** [comment]
Thanks again for the detailed guidance — the -g step -p x --col-axis y approach works exactly right. 

One small follow-up, only if it's easy: is there a way to give a series a short label for the legend (to keep it readable with many series) and a separate, longer name shown on hover/tooltip? Not a blocker, just curious if that's already possible.

**Comment by fahimfaisaal** [comment]
Yep, its not supported today. One series name is used for both the legend and the tooltip, so short `{label}`s (like `col0{A}`) is the way to keep the legend readable for now.

If we add this later, I’m leaning toward a separate map flag rather than packing two names into `{label}`:

```bash
# wide
vizb line test.csv -g step -p x --col-axis y \
  --select 'col0{A},col1{B}' \
  --series-desc 'A=Alpha train loss,B=Beta val loss'

# long (series already short codes in the data)
vizb line test.csv -g series,step -p y,x --select value \
  --series-desc 'A=Alpha train loss,B=Beta val loss'
```

That would land in the dataset roughly like:

```json
"series": [
  { "key": "A", "description": "Alpha train loss" },
  { "key": "B", "description": "Beta val loss" }
]
```

Legend / series identity stays on `key`, tooltip can show `description` when present. Not a priority yet, and there might be a better design, I’m open to hear.

**Comment by mayura-andrew** [comment]
@fahimfaisaal Hi, while working on vizb line, I found that vizb line rounds every value to 2 decimal places in its output JSON, regardless of input precision:

```
step,train_loss
1,0.914273581
2,0.913891247
3,0.912456839
4,0.910128475
5,0.909983221
```

I used this command to run,

`vizb line precision.csv -n loss -g step -p x --col-axis y --select "train_loss{Train}" -o precision.json`

Output - 

`{"timestamp":"2026-08-05T07:40:07Z","name":"loss","theme":"default","axes":[{"key":"x","label":"step"},{"key":"y"}],"settings":[{"type":"line","scale":"linear"}],"data":[{"xAxis":"1","yAxis":"Train","stats":[{"value":0.91}]},{"xAxis":"2","yAxis":"Train","stats":[{"value":0.91}]},{"xAxis":"3","yAxis":"Train","stats":[{"value":0.91}]},{"xAxis":"4","yAxis":"Train","stats":[{"value":0.91}]},{"xAxis":"5","yAxis":"Train","stats":[{"value":0.91}]}]}`

All 5 genuinely different (declining) values come out as 0.91 in every case. For metrics that are converging/plateauing — extremely common in ML training curves, our actual use case — this turns real, meaningful variation into a flat, misleading line. Is there a way to control output precision, or could there be?

**Comment by fahimfaisaal** [comment]
This is a real bug; working on it @mayura-andrew, will be fixed on the next release

**Comment by mayura-andrew** [comment]
Thanks for confirming and picking this up!

**Comment by fahimfaisaal** [comment]
Hi, @mayura-andrew, released latest `v0.18.0` install and try itshould work

**Comment by mayura-andrew** [comment]
Sure I will try, thank you :)

**Comment by mayura-andrew** [comment]
It is working fine :) Thank you!

**Comment by mayura-andrew** [comment]
Hi, another question: the line chart's X-axis always renders as categorical (evenly-spaced labels), even when every X value is a plain number (e.g. training step). Is there a way to get a continuous/numeric X-axis instead, so log-scale would also work on X? ECharts itself supports this (xAxis.type: "value" or "log") — just wasn't sure if vizb currently exposes it.

@fahimfaisaal

**Comment by fahimfaisaal** [comment]
Hi @mayura-andrew, actually I didn't find any case to support log at x axis so keept it only for y axis.

If you have the proper use case, feel free to share the proposal and how we should expose this. Thanks

**Comment by janpfeifer** [comment]
hey @fahimfaisaal ,

Let me provide a common use case example (ours): 

During training of ML models, it's most common to use the "global step" (gradient steps) on the X axis. The model improvements are often log-linear with the number of steps training. During most of the training the loss most commonly decrease linearly on the log of the number of steps -- until it saturates (other metrics that one may monitor, like accuracy, behaves similarly).

So by default I set the plots on log scale on the X axis during plots. Actually, I also only issue evaluations of the current model state at exponential steps (they are expensive to generate if one only has one machine, because generating an evaluation step pauses the training).

Training often spans to 10^3 to 10^7 steps. And we want to capture early fast improvements in the same plot.  So we also set by default the Y on logs-scale (but it depends on the plot, the Y axis is best left configurable). 

See example plot with X-log scale here (comparing image loss during training of two "flow-matching" diffusion models during training):

<img width="2748" height="922" alt="Image" src="https://github.com/user-attachments/assets/b1ecb009-5663-4c09-b01e-7807a1d77058" />

If the X-axis wasn't in log-scale, the points would be all cramped to the left of the plot, and the information wouldn't be as visible.

PS.: I should mention that both [Tensorboard](https://www.tensorflow.org/tensorboard) and [Weight&Biases](https://wandb.ai/), the most popular monitoring tools for ML training, support X-axis on log-scale monitoring of metrics.

**Comment by fahimfaisaal** [comment]
@janpfeifer Thanks for the explanation, and sorry for being late.

However, I understood. So we can bring the option for an x-axis log scale. The default will be the y-axis. and it can be configured through `--scale "log:axis=x,y;base=10"`.  By exposing it this way, I think it will be scalable.

Btw, idk how the e-chart grid will behave here. Did you find any example ECharts similar to this?

**Comment by janpfeifer** [comment]
Thanks! 

No, I didn't explore too much E-Charts ...

**Comment by fahimfaisaal** [comment]
Ok, let me allocate some time for it.

**Comment by fahimfaisaal** [comment]
@janpfeifer, I've done some experiments on the x-axis log. Do you think it's enough for the use case?

### The input
```csv
step,m1_train,m1_eval,m1_img_loss,m2_train,m2_eval,m2_img_loss
2,0.259579,0.242224,0.242144,0.225736,0.217668,0.214149
3,0.258404,0.240506,0.242067,0.220791,0.21052,0.209755
4,0.249229,0.239974,0.243501,0.221232,0.208425,0.208307
5,0.253759,0.236775,0.244364,0.220116,0.207761,0.203208
7,0.252345,0.23816,0.23429,0.213473,0.209394,0.204363
9,0.249221,0.240667,0.238481,0.216492,0.203849,0.202153
12,0.247212,0.237931,0.241095,0.211582,0.204423,0.196067
16,0.24023,0.233023,0.230424,0.207011,0.197581,0.195912
22,0.241111,0.231996,0.232485,0.204135,0.196703,0.198979
30,0.238911,0.232713,0.228714,0.204626,0.193751,0.192405
40,0.239529,0.231428,0.232065,0.201955,0.197239,0.193637
54,0.231259,0.223696,0.227836,0.19698,0.190175,0.192975
73,0.234261,0.224869,0.230582,0.195298,0.19381,0.187121
99,0.227191,0.222528,0.228109,0.192063,0.189179,0.188658
134,0.226124,0.220613,0.232012,0.191158,0.183276,0.179917
...
595728,0.174923,0.182543,0.198,0.14543,0.141167,0.140752
804233,0.175287,0.186132,0.198,0.143794,0.142631,0.139483
1085715,0.178725,0.182711,0.198,0.143588,0.144065,0.137524
1465715,0.177554,0.17759,0.198,0.142532,0.139137,0.137696
1978715,0.170923,0.179101,0.197358,0.139858,0.143114,0.137793
2671266,0.174293,0.175246,0.198,0.141102,0.143548,0.136386
3606209,0.171894,0.174502,0.194548,0.139871,0.1414,0.137339
4868382,0.173947,0.176553,0.197941,0.13524,0.1372,0.142071
```

### command

```bash
vizb line img-loss.csv -n img_loss -g step -p x -A y --smooth --scale "log:axes=x \
  --select 'm1_train{#1 train},m1_eval{#1 eval},m1_img_loss{#1 img_loss},m2_train{#2 train},m2_eval{#2 eval},m2_img_loss{#2 img_loss}'  \
  -o out.html
```

### output without log

<img width="1215" height="678" alt="Image" src="https://github.com/user-attachments/assets/eec4e8c4-1cb4-40e1-8aed-9d96ace6d962" />

### output with log (base 10 - configurable) 

<img width="1347" height="748" alt="Image" src="https://github.com/user-attachments/assets/0709a496-f426-45e0-a41d-84cd02cdd09e" />

**Comment by janpfeifer** [comment]
A couple of small issues (sorry we are pushing too hard, hopefully this will be useful for others): 

- There last point of the "#2 eval" is repeated at the end, whereas the data, the traininig session last step was  maybe ~130K. 
- Sometimes, I'm comparing training sessions where the evaluations didn't necessarily happen on the same step (X ) points. Meaning the values for X of the different series may be different. Does e-chart accept that ?

**Comment by fahimfaisaal** [comment]
> There last point of the "https://github.com/goptics/vizb/pull/2 eval" is repeated at the end, whereas the data, the traininig session last step was maybe ~130K.

I didn't provide the full input sample there, and I think the repetition you mentioned isn't `#2 eval` its `#1 img_loss` which has multiple  `0.194` approx 8-10 at the y-axis.

> Sometimes, I'm comparing training sessions where the evaluations didn't necessarily happen on the same step (X ) points. Meaning the values for X of the different series may be different. Does e-chart accept that ?

If I get you properly, are you mentioning about `missing value`? like this

<img width="1355" height="764" alt="Image" src="https://github.com/user-attachments/assets/b2c83275-dfe1-4e4e-9b93-b6f5710191c2" />

**Comment by janpfeifer** [comment]
> If I get you properly, are you mentioning about `missing value`? like this

Yes, that's exactly it! Thanks!

**Comment by fahimfaisaal** [comment]
Ok, then Vizb already supports the missing value. Just need to pass the column value as empty.

And at the next release, the x-log feature will be there.

Thanks for being with Vizb.

**Comment by fahimfaisaal** [comment]
hi @janpfeifer, `v0.20.0` has been released with an extended log feature.  just run `vizb udpate` to get the updates.

btw we also have a slider data zoom feature which unlocks on when the X series length more than `50`

https://github.com/user-attachments/assets/05898840-60f8-4b43-835a-ba0eaa536607
